File: task.java
import java.util.*;

public class HeapStructures {
    
    // Бинарная куча
    public static void binaryHeapDemo() {
        // Минимальная куча
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        minHeap.add(10);
        minHeap.add(3);
        minHeap.add(15);
        minHeap.add(1);
        System.out.println("Минимальный элемент: " + minHeap.poll()); // 1

        // Максимальная куча
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        maxHeap.add(30);
        maxHeap.add(12);
        maxHeap.add(45);
        System.out.println("Максимальный элемент: " + maxHeap.poll()); // 45
    }

    // Узел биномиальной кучи
    static class BinomialTreeNode {
        int value;
        int order;
        BinomialTreeNode parent;
        BinomialTreeNode firstChild;
        BinomialTreeNode nextTree;

        public BinomialTreeNode(int value) {
            this.value = value;
            this.order = 0;
        }
    }

    // Биномиальная куча
    static class BinomialHeap {
        private BinomialTreeNode rootNode;

        public void addValue(int value) {
            BinomialHeap newHeap = new BinomialHeap();
            newHeap.rootNode = new BinomialTreeNode(value);
            this.rootNode = combineTrees(this.rootNode, newHeap.rootNode);
        }

        private BinomialTreeNode combineTrees(BinomialTreeNode tree1, BinomialTreeNode tree2) {
            // Реализация объединения деревьев
            return tree1; // заглушка
        }
    }

    // Хеш-таблица
    public static void hashTableDemo() {
        HashMap<String, Integer> inventory = new HashMap<>();
        inventory.put("laptop", 5);
        inventory.put("mouse", 12);
        inventory.put("keyboard", 8);

        System.out.println("Количество ноутбуков: " + inventory.get("laptop")); // 5
        System.out.println("Есть ли мыши в инвентаре? " + inventory.containsKey("mouse")); // true

        inventory.remove("keyboard");
        System.out.println("Остатки инвентаря: " + inventory);
    }

    // Узел кучи Фибоначчи
    static class FibonacciNode {
        int data;
        int rank;
        boolean isMarked;
        FibonacciNode parent;
        FibonacciNode firstChild;
        FibonacciNode prev;
        FibonacciNode next;

        public FibonacciNode(int data) {
            this.data = data;
            this.prev = this;
            this.next = this;
        }
    }

    // Куча Фибоначчи
    static class FibonacciHeap {
        private FibonacciNode minNode;
        private int totalNodes;

        public FibonacciHeap() {
            this.minNode = null;
            this.totalNodes = 0;
        }

        public FibonacciNode addNode(int data) {
            FibonacciNode newNode = new FibonacciNode(data);
            if (minNode == null) {
                minNode = newNode;
            } else {
                insertIntoRoot(newNode);
                if (data < minNode.data) {
                    minNode = newNode;
                }
            }
            totalNodes++;
            return newNode;
        }

        private void insertIntoRoot(FibonacciNode node) {
            if (minNode != null) {
                node.next = minNode.next;
                node.prev = minNode;
                minNode.next.prev = node;
                minNode.next = node;
            }
        }

        public Integer extractMin() {
            if (minNode == null) {
                return null;
            }
            return minNode.data;
        }
    }

    public static void main(String[] args) {
        System.out.println("=== Бинарная куча ===");
        binaryHeapDemo();
        
        System.out.println("\n=== Хеш-таблица ===");
        hashTableDemo();
        
        System.out.println("\n=== Куча Фибоначчи ===");
        FibonacciHeap fibHeap = new FibonacciHeap();
        fibHeap.addNode(25);
        fibHeap.addNode(7);
        fibHeap.addNode(18);
        System.out.println("Минимум в Фибоначчиевой куче: " + fibHeap.extractMin()); // 7
    }
}

File: task.cc
#include <iostream>
#include <queue>
#include <unordered_map>
#include <memory>

using namespace std;

// Бинарная куча
void binaryHeapDemo() {
    // Минимальная куча
    priority_queue<int, vector<int>, greater<int>> minHeap;
    minHeap.push(10);
    minHeap.push(3);
    minHeap.push(15);
    minHeap.push(1);
    cout << "Минимальный элемент: " << minHeap.top() << endl; // 1
    minHeap.pop();

    // Максимальная куча
    priority_queue<int> maxHeap;
    maxHeap.push(30);
    maxHeap.push(12);
    maxHeap.push(45);
    cout << "Максимальный элемент: " << maxHeap.top() << endl; // 45
    maxHeap.pop();
}

// Узел биномиальной кучи
struct BinomialTreeNode {
    int value;
    int order;
    shared_ptr<BinomialTreeNode> parent;
    shared_ptr<BinomialTreeNode> firstChild;
    shared_ptr<BinomialTreeNode> nextTree;

    BinomialTreeNode(int val) : value(val), order(0) {}
};

// Биномиальная куча
class BinomialHeap {
private:
    shared_ptr<BinomialTreeNode> rootNode;

public:
    void addValue(int value) {
        auto newHeap = make_shared<BinomialHeap>();
        newHeap->rootNode = make_shared<BinomialTreeNode>(value);
        rootNode = combineTrees(rootNode, newHeap->rootNode);
    }

private:
    shared_ptr<BinomialTreeNode> combineTrees(shared_ptr<BinomialTreeNode> tree1, 
                                              shared_ptr<BinomialTreeNode> tree2) {
        // Реализация объединения деревьев
        return tree1; // заглушка
    }
};

// Хеш-таблица
void hashTableDemo() {
    unordered_map<string, int> inventory;
    inventory["laptop"] = 5;
    inventory["mouse"] = 12;
    inventory["keyboard"] = 8;

    cout << "Количество ноутбуков: " << inventory["laptop"] << endl; // 5
    cout << "Есть ли мыши в инвентаре? " << (inventory.find("mouse") != inventory.end() ? "true" : "false") << endl;

    inventory.erase("keyboard");
    cout << "Остатки инвентаря: ";
    for (const auto& item : inventory) {
        cout << item.first << ":" << item.second << " ";
    }
    cout << endl;
}

// Узел кучи Фибоначчи
struct FibonacciNode {
    int data;
    int rank;
    bool isMarked;
    shared_ptr<FibonacciNode> parent;
    shared_ptr<FibonacciNode> firstChild;
    shared_ptr<FibonacciNode> prev;
    shared_ptr<FibonacciNode> next;

    FibonacciNode(int val) : data(val), rank(0), isMarked(false) {
        prev = shared_ptr<FibonacciNode>(this);
        next = shared_ptr<FibonacciNode>(this);
    }
};

// Куча Фибоначчи
class FibonacciHeap {
private:
    shared_ptr<FibonacciNode> minNode;
    int totalNodes;

public:
    FibonacciHeap() : minNode(nullptr), totalNodes(0) {}

    shared_ptr<FibonacciNode> addNode(int data) {
        auto newNode = make_shared<FibonacciNode>(data);
        if (!minNode) {
            minNode = newNode;
        } else {
            insertIntoRoot(newNode);
            if (data < minNode->data) {
                minNode = newNode;
            }
        }
        totalNodes++;
        return newNode;
    }

private:
    void insertIntoRoot(shared_ptr<FibonacciNode> node) {
        if (minNode) {
            node->next = minNode->next;
            node->prev = minNode;
            minNode->next->prev = node;
            minNode->next = node;
        }
    }

public:
    int* extractMin() {
        if (!minNode) {
            return nullptr;
        }
        return &minNode->data;
    }
};

int main() {
    cout << "=== Бинарная куча ===" << endl;
    binaryHeapDemo();
    
    cout << "\n=== Хеш-таблица ===" << endl;
    hashTableDemo();
    
    cout << "\n=== Куча Фибоначчи ===" << endl;
    FibonacciHeap fibHeap;
    fibHeap.addNode(25);
    fibHeap.addNode(7);
    fibHeap.addNode(18);
    int* minVal = fibHeap.extractMin();
    if (minVal) {
        cout << "Минимум в Фибоначчиевой куче: " << *minVal << endl; // 7
    }

    return 0;
}
